File: RoboLabWP/RoboLabWP/Model/Game/MZMaze.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using RoboLabWP.Model.Basic_Types;

using System.Windows.Resources;
using System.Windows;

//
//  MZMaze.cs
//  Maze
//
//  Класс лабиринта. Хранит данные о его структуре. Не изменяется на протяжении жизни объекта.

namespace RoboLabWP.Model.Game
{
    public class MZMaze
    {
        int _width;
        int _height;
        int _minimalStepsRequired;
        
        MZPosition _winningPosition;
        MZPosition _startingPosition;
        MZCellMatrix _cellMatrix;
        List<MZFlow> _flowList;

        public MZMaze()
        {

            int cuurent_mod = (Application.Current as App).UserGameData.current_module;
            string path = (Application.Current as App).UserGameData.GameSettings.what_module(cuurent_mod);
            path = string.Concat(path,".mzp");
            //string path = "kal.mzp";            
            Uri uri = new Uri(path, UriKind.Relative);
            
            StreamResourceInfo sri = Application.GetResourceStream(uri);
            byte[] data = new byte[sri.Stream.Length];
            sri.Stream.Read(data, 0, data.Length);
            
            //int level = 1;
            int level = (Application.Current as App).UserGameData.levelnumber;
            
            int n = data[0];
            int baseOffset = 1 + n * 2 + data[1 + (level - 1) * 2] * 256 + data[2 + (level - 1) * 2]; 

            _width = data[baseOffset];
            _height = data[baseOffset + 1];
            _minimalStepsRequired = data[baseOffset + 2] * 256 + data[baseOffset + 3];
            _cellMatrix = new MZCellMatrix(_width, _height);

            for (int i = 0; i < _width; i++)
            {
                for (int j = 0; j < _height; j++)
                {
                    MZPosition thisPosition = new MZPosition(i, j);
                    MZCell thisCell = _cellMatrix.cellAtPosition(thisPosition);
                    MZByte thisCellData = new MZByte(data[baseOffset + i * _height + j + 4]);
                    MZByte walls = new MZByte((byte)(thisCellData.value() & MZMacro.LOW4BIT));
                    bool isWin = (thisCellData.value() & MZMacro.BIT5) != 0;
                    bool isStart = (thisCellData.value() & MZMacro.BIT6) != 0;
                    bool isEmpty = (thisCellData.value() & MZMacro.BIT7) != 0;
                    thisCell.isEmpty = isEmpty;
                    thisCell.setWallsValue(walls);
                    if (isWin)
                        _winningPosition = thisPosition;
                    if (isStart)
                        _startingPosition = thisPosition;
                }
            }
            _flowList = new List<MZFlow>();

            /*
            if(userGameData currentModuleName] isEqualToString: @"Reciever"] && level == 2)
            {
               [_flowList addObject:[[MZFlow alloc] initWithPosition: [[MZPosition alloc] initWithX:4 andY:8] direction: RIGHT length:6 andSpeed: 1.0/15.0]];
            }*/
            //cuurent_mod == 4 - Reciever
            if ((cuurent_mod == 4) && (level == 2))
            {
                MZPosition pos = new MZPosition(4, 8);
                MZFlow flow = new MZFlow(pos, MZDirection.RIGHT,6,1.0/15.0);
                _flowList.Add(flow);
            }
        }
        public MZCellMatrix cells()
        {
            return _cellMatrix;
        }

        public MZPosition startingPosition()
        {
            return _startingPosition;
        }

        public MZPosition winningPosition()
        {
            return _winningPosition;
        }

        public int width()
        {
            return _width;
        }

        public int height()
        {
            return _height;
        }

        public int minimalSteps()
        {
            return _minimalStepsRequired;
        }
        public List<MZFlow> flowList()
        {
            return _flowList;
        }
    }
}

File: RoboLabWP/RoboLabWPComp/Model/Game/MZMaze.cpp
#include "pch.h"
#include "MZMaze.h"


MZMaze::MZMaze(void)
{
    //MZUserGameData *userGameData = [MZUserGameData getInstance];
    //NSMutableData * moduleData= [userGameData currentModuleData];
    //MZByte* byteArray = [moduleData mutableBytes];
    //int level = [userGameData currentLevel];
	
	MZByte* byteArray = new MZByte(100); // временно
	int level = 1;
    
	int n = byteArray[0];
    int baseOffset = 1 + n * 2 + byteArray[1 + (level - 1) * 2] * 256 + byteArray[2 + (level - 1) * 2]; //вникать не надо. В это можно просто поверить.
	MZPosition* winningPosition = NULL;
	_startingPosition = NULL;

    _width = byteArray[baseOffset];
    _height = byteArray[baseOffset + 1];

	_cellMatrix = new MZCellMatrix(_width, _height);
    _minimalStepsRequired = byteArray[baseOffset + 2] * 256 + byteArray[baseOffset + 3];
    
	for(int i = 0; i < _width; i++)
    {
        for(int j = 0; j < _height; j++)
        {
			MZPosition *thisPosition  = new MZPosition(i,j);
			MZCell *thisCell = _cellMatrix->cellAtPosition(thisPosition);
            MZByte thisCellData = byteArray[baseOffset + i*_height + j + 4];
            MZByte walls = thisCellData & LOW4BIT;
            bool isWin = thisCellData & BIT5;
            bool isStart = thisCellData & BIT6;
            bool isEmpty = thisCellData & BIT7;
			thisCell->isEmpty = isEmpty;
			thisCell->setWallsValue(walls);
            if(isWin)
            {
                winningPosition = thisPosition;
            }
            if(isStart)
            {
                _startingPosition = thisPosition;
            }
        }
    }
    
    //_objectList = [[NSMutableArray alloc] init];
    //[_objectList addObject:[[MZExit alloc] initWithPosition:winningPosition]];
	_objectList.push_back(new MZExit(winningPosition));
	delete(winningPosition);
	//string currentmodule = "Reciever";
	int currentModuleNumber = 2;
	MZPosition *pos;
	//if([[userGameData currentModuleName] isEqualToString: @"Reciever"] && level == 1)
    if(currentModuleNumber == 2 && level == 2)
    {
        //MZFlow *flow = [[MZFlow alloc] initWithPosition: [[MZPosition alloc] initWithX:4 andY:8] direction: RIGHT length:6 andSpeed: 1.0/15.0];
		pos = new MZPosition(4,8);
		MZFlow *flow = new MZFlow(pos, RIGHT, 6, 1.0/15.0);
		delete(pos);
        //[_objectList addObject:flow];
        //[_objectList addObject:[[MZButton alloc] initWithPosition: [[MZPosition alloc] initWithX:10 andY:7] andDirection:DOWN andTarget:flow]];
		_objectList.push_back(flow);
		pos = new MZPosition(10,7);
		_objectList.push_back(new MZButton(pos, DOWN, flow));
		delete(pos);
    }
    if(currentModuleNumber == 2 && level == 1)
    {
        //[_objectList addObject:[[MZOneWayWall alloc] initWithPosition:[[MZPosition alloc] initWithX:8 andY: 5] andDirection:LEFT]];
		pos = new MZPosition(8,5);
		_objectList.push_back(new MZOneWayWall(pos, LEFT));
		delete(pos);
    }
    if(currentModuleNumber == 2 && level == 3)
    {
		MZCell* cell = new MZCell((MZByte)1);
		pos = new MZPosition(2,12);
        //MZRotatingCell* rotatingCell = [[MZRotatingCell alloc] initWithPosition:[[MZPosition alloc] initWithX:2 andY:12] andWalls:cell andRotationDirection:CLOCKWISE];
		MZRotatingCell* rotatingCell = new MZRotatingCell(pos, cell, CLOCKWISE);
		delete(pos);
        //[_objectList addObject:rotatingCell];
		_objectList.push_back(rotatingCell);
        //[_objectList addObject:[[MZButton alloc] initWithPosition:[[MZPosition alloc] initWithX:0 andY:12] andDirection:UP andTarget:rotatingCell]];
		pos = new MZPosition(0,12);
		_objectList.push_back(new MZButton(pos, UP, rotatingCell));
		delete(pos);
    }
}


MZMaze::~MZMaze(void)
{
	if(_cellMatrix != NULL)
	{
		delete(_cellMatrix);
		_cellMatrix = NULL;
	}
	if(_startingPosition != NULL)
	{
		delete(_startingPosition);
		_startingPosition = NULL;
	}
	for (unsigned int i = 0; i < _objectList.size(); i++)
	{
		if (_objectList[i] != NULL)
		{
			delete(_objectList[i]);
			_objectList[i] = NULL;
		}
	}
}

MZPosition* MZMaze::startingPosition()
{
    return _startingPosition;
}
MZCellMatrix* MZMaze::cells()
{
    return _cellMatrix;
}
vector<MZGameObject*> MZMaze::objectList()
{
	return _objectList;
}
int MZMaze::width()
{
    return _width;
}
int MZMaze::height()
{
    return _height;		
}
int MZMaze::minimalSteps()
{
    return _minimalStepsRequired;
}
